README.md

```diff
@@ -1 +1,649 @@
+
+| Term      | Meaning                                                          |
+| --------- | ---------------------------------------------------------------- |
+| Attacker  | A malicious third party attempting to gain access to the channel |
+| ECDH      | Elliptic Curve Diffie-Hellman                                    |
+| PK        | Public key                                                       |
+| Requestor | The agent opening the session                                    |
+| Responder | The agent being contacted by the Requestor                       |
+| SK        | Secret (private) key                                             |
+
+## 1.2 Payload Fields
+
+Payloads are encoding agnostic, but JSON is RECOMMENDED. For JSON, any fields that contain non-JSON values (such as ECDH public keys and encryption payloads) MUST be serialized as unpadded [Base64](https://datatracker.ietf.org/doc/html/rfc4648).
+
+Messages that a peer cannot parse SHOULD be ignored.
+
+All payloads MUST include the "AWAKE version" field `awv: "0.1.0"`. Payloads MUST also include a message type field `type` (see each stage for the value). All field keys and message type values MUST be lowercase and treated as case-sensitive.
+
+| Field  | Value          | Description           | Required |
+| ------ | -------------- | --------------------- | -------- |
+| `awv`  | `"0.1.0"`      | AWAKE message version | Yes      |
+| `type` | `awake/<type>` | AWAKE message type    | Yes      |
+
+## 1.4 Encryption
+
+Encryption is core to securing a tunnel. Key material and secrets created for AWAKE MUST be considered ephemeral and MUST NOT be reused between sessions.
+
+At a high-level, AWAKE uses a NIST P-256 [Elliptic Curve Diffie-Hellman](https://en.wikipedia.org/wiki/Elliptic-curve_Diffie%E2%80%93Hellman) (ECDH) [Double Ratchet](https://signal.org/docs/specifications/doubleratchet/) to secure messages.
+
+### 1.4.1 Asymmetric Keys
+
+UCAN MUST be used as the signature envelope for AWAKE. Any UCAN-compatible asymmetric key MAY be used for signatures, including RSA, Ed25519, P-256, and so on.
+
+The [ECDH portion](#1431-ecdh-input) of the [Double Ratchet KDF](#143-key-derivation) MUST use P-256.
+
+### 1.4.2 Symmetric Keys
+
+All symmetric encryption in AWAKE MUST use [256-bit AES-GCM](https://csrc.nist.gov/publications/detail/sp/800-38d/final). These keys MUST be generated by the [KDF](#143-key-derivation), and SHOULD be non-extractable where possible.
+
+Each encrypted payload MUST include a unique (freshly generated) 12-byte [initialization vector](https://en.wikipedia.org/wiki/Initialization_vector).
+
+### 1.4.3 Key Derivation
+
+AWAKE uses a modified [Extended Triple Diffie-Hellman (X3DH)](https://www.signal.org/docs/specifications/x3dh/) protocol in order to support the WebCrypto API's non-extractable keys and make use of existing UCAN tokens without the need for pre-sharing keys.
+
+Key derivation in AWAKE's double ratchet MUST use the following algorithm:
```

Comment:
Is there a reason you're using SHA3 to derive the OKM directly, rather than going through something like [HKDF](https://developer.mozilla.org/en-US/docs/Web/API/HkdfParams)?

I notice that the spec doesn't give any guidance on generating the IV, but with HKDF you'd be able to generate both the OKM and the IV from the output of `ECDH`, and avoid needing to send it along through the protocol. You'd just need to make sure to use the `info` parameter of HKDF to guarantee domain separation for both derivations.

Reply:
HKDF was used in an earlier draft, because it's available in the WebCrypto API. The short answer is that we can totally use it if it's better. You've gone deeper on this stuff than me, so I'd love to explain the reasoning and get your thoughts!

# What Does Signal Do?

Signal's Double Ratchet paper suggests two systems for the KDF: HMAC (SHA) and HKDF.

# In Favour of HKDF

HKDF intends to be an easy choice to "just use" as a generic KDF. The people that designed it are -- frankly -- much better at cryptography than me!

# Against HKDF

On the flip side, hash chaining HMACs to derive keys is pretty widely used.

HKDF's security comes from SHA2 which is significantly slower than SHA3. HKDF is much more general than a hash function, and does things like expansion:

> 2. To "expand" the generated output of an already reasonably random input such as an existing shared key into a larger cryptographically independent output
> — Wikipedia

In our use case, we need exactly 256-bit outputs, so features like expansion (and the additional computation therein) isn't required. Breaking a SHA3-derived key is the same as reversing the hash, which "should" be intractable (under hash assumptions).

Reply:
> You've gone deeper on this stuff than me, so I'd love to explain the reasoning and get your thoughts!

This also definitely isn't my expertise, and most of what I know is just sourced from cryptographers with far more experience than me!

I don't believe there's anything insecure about using SHA3 here, like you are, but I've got two comments.

> HKDF's security comes from SHA-2 which is significantly slower than SHA-3. 

This sounds backwards to me? I thought SHA-3 was widely considered to be slower, except on specialized hardware. This is a point that's brought up by [Adam Langley](https://www.imperialviolet.org/2017/05/31/skipsha3.html):

> SHA-3 is also [slow](http://bench.cr.yp.to/results-sha3.html), and is even slower than SHA-2 which is already a comparative laggard amongst crypto primitives.

And is also something [David Wong acknowledges](https://cryptologie.net/article/400/maybe-you-shouldnt-skip-sha-3/). He concludes that you should just choose SHA-3 as a general-purpose hashing algorithm anyway, because avoiding the risk of length extension attacks are worth the slowdown, but that's not something we need to worry about with HKDF.

I also [found some hashcat benchmarks](https://gist.github.com/Chick3nman/ccfb883d2d267d94770869b09f5b96ed) someone ran on an M1 earlier this year, and it looks like SHA-2 outperforms SHA-3 there:

```
---------------------------
* Hash-Mode 1400 (SHA2-256)
---------------------------

Speed.#1.........:  1818.9 MH/s (54.07ms) @ Accel:128 Loops:128 Thr:128 Vec:1

----------------------------
* Hash-Mode 17400 (SHA3-256)
----------------------------

Speed.#1.........:   581.7 MH/s (85.14ms) @ Accel:128 Loops:64 Thr:128 Vec:1
```

> In our use case, we need exactly 256-bit outputs, so features like expansion (and the additional computation therein) isn't required.

I'd actually argue that we need 352-bit outputs, because we have a 256-bit key, and a 96-bit IV. Whether or not it's generated as part of the KDF, that IV is needed at some point, so implementers will need to find another 12-bytes each round from somewhere, and I'd argue that baking the generation of the IV into the KDF would both limit possible mistakes elsewhere (using the same IV in both directions, for example).

> I'm weary of pseudorandom IVs, but maybe that's just me?

My understanding of the IV in AES-GCM is that it just needs to be unique, so one of my concerns is that someone will attempt to do so by either using a hardcoded IV (because each party only sends a single message per key), or by using a counter. In both cases they'll be forgetting to take into account that the other party also sends a message, and so would be making use of the same IV.

My thought with using HKDF to derive the IV was to use the sender's DID as the salt to HKDF so that both parties are able to use the same key material to generate a unique IV per round.

> There's a few places in AWAKE that we could be more economical with space, but ease of implementation is the bigger concern for this version.

That makes sense! I agree that saving space isn't the highest priority here, and the larger point behind the suggestion was about removing a degree of freedom from implementers that may lead to IV misuse, with removing that field just being a bonus.

Either way though, this comment definitely isn't me saying that you're doing anything wrong and that it needs to change or anything like that! I was mostly curious about what had led you toward SHA-3 in the first place.

Reply:
> This sounds backwards to me? I thought SHA-3 was widely considered to be slower, except on specialized hardware

We ran some benchmarks roughly last autumn to test out Wasm, and IIRC the Wasm SHA3-256 outperformed the WebCrypto API's SHA2-256 (not an apples-to-apples comparison, but you'd expect the native function to outperform Wasm). IIRC the SHA3 team also believes this to be the case for pure software implementations.

But that said, we control the stack here, so SHA2 or BLAKE3 or similar are all on the table!

Lemme dig up that post about SHA2 vs SHA3, though.

Reply:
> My understanding of the IV in AES-GCM is that it just needs to be unique

Okay, yeah, that's true!

> In both cases they'll be forgetting to take into account that the other party also sends a message, and so would be making use of the same IV.

That is a _really good point_ 💯

Reply:
> We ran some benchmarks roughly last autumn to test out Wasm, and IIRC the Wasm SHA3-256 outperformed the WebCrypto API's SHA2-256

Fascinating! I'm seeing the same thing echoed in other benchmarks, like [this](https://github.com/bitauth/libauth/wiki/Benchmarks#browser-sha256-hash-a-32-byte-input) `libauth` benchmark that puts their Wasm sha256 implementation at almost 20x the throughput of `crypto.subtle`, for 32-byte inputs:

```
libauth x 571,012 ops/sec ±2.38% (62 runs sampled)
hash.js x 346,025 ops/sec ±1.31% (63 runs sampled)
crypto.subtle x 37,260 ops/sec ±2.15% (58 runs sampled)
asmcrypto.js x 3,138 ops/sec ±2.38% (20 runs sampled)
```